web-app/HeimScripts/heatmap/run.R

```diff
@@ -393,3 +507,43 @@ addClusteringOutput <- function(jsn, measurements) {
   )
   return(jsn)
 }
+
+### Limma part. It will be moved into a separate
+# module when we get R scripts sourcing to work here.
+
+getDEgenes <- function(df) {
+  measurements    <- getMeasurements(df)
+  enoughSamples   <- ncol(measurements) > 3 && ncol(measurements) - getSubset1Length(measurements) > 1
+  if (!enoughSamples){
+    stop("Not enough samples to find differentially expressed genes.")
+  }
+  design          <- getDesign(measurements)
+  contrast.matrix <- makeContrasts( S1-S2, levels = design )
+  fit             <- lmFit(measurements, design)
+  fit             <- contrasts.fit(fit, contrast.matrix)
+  fit             <- eBayes(fit)
+  contr           <- 1  #  We need a vector, not a df, so we'll do [, contr] on all stats
+  top.fit         <- data.frame (
+    logfold = fit$coefficients[, contr],
+    ttest   = fit$t[, contr],
+    pval    = fit$p.value[, contr],
+    adjpval = p.adjust(
+      p      = fit$p.value[, contr],
+      method ='fdr'),
+    bval    = fit$lods[, contr]
+  )
+  cbind(df[,1:2], top.fit)
+}
+
+getDesign <- function(measurements) {
+  subset1Length <- getSubset1Length(measurements)
+  classVectorS1 <- c(rep(1, subset1Length), rep(2, ncol(measurements) - subset1Length ))
+  classVectorS2 <- rev(classVectorS1)
```

Comment:
Should the classVectorS1 and classVectorS2 be exactly opposites to each other? 
Right now that is only the case if S1 is of the same size as S2. If that's not the case the vectors are not exactly opposite, example:




df
  GSM210004_n0_s1 GSM210005_n0_s1 GSM210006_n1_s1 GSM210007_n1_s1 GSM210010_n0_s1 GSM210005_n0_s2
1       179.26000       185.84100       179.15000       367.43600       180.26000       186.84100
2              NA        42.23340        35.58670        60.10030              NA        43.23340
3         0.00000        16.79990        33.93710        31.39620         1.00000        17.79990
4         0.00734        64.56860       105.91200       140.32500         1.00734        65.56860
5       -23.02340         3.82495         1.15958         5.08933       -22.02340         4.82495
  GSM210006_n1_s2 GSM210007_n1_s2
1       180.15000       368.43600
2        36.58670        61.10030
3        34.93710        32.39620
4       106.91200       141.32500
5         2.15958         6.08933


design <- getDesign(df)
> design
     S1 S2
[1,]  1  2
[2,]  1  2
[3,]  1  2
[4,]  1  1
[5,]  1  1
[6,]  2  1
[7,]  2  1
[8,]  2  1